src/sage/geometry/cone_critical_angles.py

```diff
@@ -0,0 +1,998 @@
+
+
+def _solve_gevp_naive(GG, HH, M, I, J):
+    r"""
+    Solve the generalized eigenvalue problem in Theorem 3
+    [Or2020]_ in a very naive way, by (slowly) inverting the matrices
+    and finding the eigenvalues in the product space.
+
+    This is used only for testing, to ensure that the smart way of
+    solving the generalized eigenvalue problem via
+    :func:`solve_gevp_zero` and :func:`solve_gevp_nonzero` returns the
+    same answers as the dumb way.
+
+    This returns a generator, like :func:`solve_gevp_zero` and
+    :func:`solve_gevp_nonzero`.
+
+    INPUT:
+
+    See the arguments for :func:`solve_gevp_nonzero`.
+
+    ALGORITHM:
+
+    We construct the two matrices `A` and `B` in Theorem 3 in block
+    form, and then use the naive "inverse" method on `B` to move it to
+    the left and obtain `M = B^{-1}A`. We then compute the right
+    eigenvectors of the whole big matrix `M`.
+
+    EXAMPLES:
+
+    A simple usage example, that also appears as Example 3 in
+    [Or2020]_::
+
+        sage: from sage.geometry.cone_critical_angles import _solve_gevp_naive
+        sage: K = cones.nonnegative_orthant(2)
+        sage: G = matrix.column(K.rays())
+        sage: GG = G.transpose()*G
+        sage: I = [0]
+        sage: J = [1]
+        sage: list(_solve_gevp_naive(GG,GG,GG,I,J))
+        [(0, (1), (0), 2), (0, (0), (1), 2)]
+
+    Check Example 4 [Or2020]_ symbolically to ensure that we get
+    eigenspaces of dimension `n=2` corresponding to the eigenvalues
+    `\cos\theta = -1` and `\cos\theta = 1`::
+
+        sage: from sage.geometry.cone_critical_angles import _solve_gevp_naive
+        sage: g11,g12,g21,g22 = SR.var('g11,g12,g21,g22', domain='real')
+        sage: h11,h12,h21,h22 = SR.var('h11,h12,h21,h22', domain='real')
+        sage: gs = [[g11,g12], [g21,g22]]
+        sage: hs = [[h11,h12], [h21,h22]]
+        sage: G = matrix.column(gs)
+        sage: H = matrix.column(hs)
+        sage: GG = G.transpose()*G
+        sage: HH = H.transpose()*H
+        sage: M = G.transpose()*H
+        sage: I = [0, 1]
+        sage: J = [0, 1]
+        sage: all( v in [-1,1] and m == 2
+        ....:      for (v,_,_,m) in _solve_gevp_naive(GG,HH,M,I,J) )
+        True
+    """
+    A = matrix.block([ [ ZZ(0),              M[I,J] ],
+                       [ M.transpose()[J,I], ZZ(0)  ] ])
+    B = matrix.block([ [ GG[I,I], ZZ(0)   ],
+                       [ ZZ(0),   HH[J,J] ] ])
+    M = B.inverse() * A
+
+    # We'll format the result to match the solve_gevp_nonzero() return value.
+    for (evalue, evectors, multiplicity) in M.eigenvectors_right():
+        for z in evectors:
+            xi = z[0:len(I)]
+            xi.set_immutable()
+            eta = z[len(I):]
+            eta.set_immutable()
+            yield (evalue, xi, eta, multiplicity)
+
+
+def solve_gevp_zero(M, I, J):
+    r"""
+    Solve the generalized eigenvalue problem in Theorem 3 for a zero
```

Comment:
```suggestion
    Solve the generalized eigenvalue problem in Theorem 3 [Or2020]_ for a zero
```
Otherwise, which Theorem 3?

Reply:
Done